solidity/contracts/BancorNetwork.sol

```diff
@@ -450,6 +405,191 @@ contract BancorNetwork is IBancorNetwork, TokenHolder, ContractRegistryClient, F
     }
 
+    /**
+      * @dev validates msg.value and prepares the conversion source token for the conversion
+      * 
+      * @param _sourceToken source token of the first conversion step
+      * @param _smartToken  smart token of the first conversion step
+      * @param _amount      amount to convert from, in the source token
+    */
+    function handleSourceToken(IERC20Token _sourceToken, ISmartToken _smartToken, uint256 _amount) private {
+        bool isSourceETH = etherTokens[_sourceToken];
+
+        // validate msg.value
+        if (msg.value > 0)
+            require(msg.value == _amount && isSourceETH);
+        else
+            require(msg.value == 0);
+
+        IBancorConverter firstConverter = IBancorConverter(_smartToken.owner());
+        bool isNewerConverter = isV27OrHigherConverter(firstConverter);
+
+        // ETH / EtherToken
+        if (isSourceETH) {
+            // handle ETH
+            if (msg.value > 0) {
+                // EtherToken converter - deposit the ETH into the EtherToken
+                // note that it can still be a non ETH converter if the path is wrong
+                // but such conversion will simply revert
+                if (!isNewerConverter) {
+                    IEtherToken etherToken = getConverterEtherTokenAddress(firstConverter);
+                    IEtherToken(etherToken).deposit.value(msg.value)();
+                }
+            }
+            // handle EtherToken
+            else {
+                // claim the tokens
+                safeTransferFrom(_sourceToken, msg.sender, this, _amount);
+
+                // ETH converter - withdraw the ETH
+                if (isNewerConverter)
+                    IEtherToken(_sourceToken).withdraw(_amount);
+            }
+        }
+        // other ERC20 token
+        else {
+            // newer converter - transfer the tokens from the sender directly to the converter
+            // otherwise claim the tokens
+            if (isNewerConverter)
+                safeTransferFrom(_sourceToken, msg.sender, firstConverter, _amount);
+            else
+                safeTransferFrom(_sourceToken, msg.sender, this, _amount);
+        }
+    }
+
+    /**
+      * @dev handles the conversion target token if the network still holds it at the end of the conversion
+      * 
+      * @param _data        conversion data, see ConversionStep struct above
+      * @param _amount      conversion return amount, in the target token
+      * @param _beneficiary wallet to receive the conversion result
+    */
+    function handleTargetToken(ConversionStep[] _data, uint256 _amount, address _beneficiary) private {
+        ConversionStep memory stepData = _data[_data.length - 1];
+
+        // network contract doesn't hold the tokens, do nothing
+        if (stepData.beneficiary != address(this))
+            return;
+
+        IERC20Token targetToken = stepData.targetToken;
+
+        // ETH / EtherToken

```

Comment:
I don't think that ETH is properly supported here (as the comment `ETH / EtherToken` implies).
I'm assuming that you're counting on `if (etherTokens[targetToken])` to hold true when `targetToken == address(0)`.
Is that correct?
If yes, then calling `IEtherToken(targetToken).withdrawTo(...)` will revert.
If no, then calling `safeTransfer(targetToken, ...)` will revert.

Reply:
in `createConversionData`, we replace whatever the user passed in the path with the correct "ether token" representation, based on the conversion version:

```
if (stepData.isV27OrHigherConverter)
    stepData.targetToken = IERC20Token(0);
    // older converter - replace the source token with the EtherToken address used by the converter
    else
        stepData.targetToken = getConverterEtherTokenAddress(stepData.converter);
```

So for an old (EtherToken) converter, `targetToken` will actually be the EtherToken used by the specific converter.